coderd/httpmw/realip.go

```diff
@@ -0,0 +1,310 @@
+package httpmw
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"strings"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/coderd/httpapi"
+)
+
+const (
+	// Note: these should be canonicalized (see http.CanonicalHeaderKey)
+	// or else things will not work correctly
+	headerCFConnectingIP  string = "Cf-Connecting-Ip"
+	headerTrueClientIP    string = "True-Client-Ip"
+	headerXRealIP         string = "X-Real-Ip"
+	headerXForwardedFor   string = "X-Forwarded-For"
+	headerXForwardedProto string = "X-Forwarded-Proto"
+)
+
+var headersAll = []string{
+	headerCFConnectingIP,
+	headerTrueClientIP,
+	headerXRealIP,
+	headerXForwardedFor,
+	headerXForwardedProto,
+}
+
+// RealIPConfig configures the search order for the function, which controls
+// which headers to consider trusted.
+type RealIPConfig struct {
+	// TrustedOrigins is a list of networks that will be trusted. If
+	// any non-trusted address supplies these headers, they will be
+	// ignored.
+	TrustedOrigins []*net.IPNet
+
+	// CloudflareConnectingIP trusts the CF-Connecting-IP header.
+	// https://support.cloudflare.com/hc/en-us/articles/206776727-Understanding-the-True-Client-IP-Header
+	CloudflareConnectingIP bool
+
+	// TrueClientIP trusts the True-Client-IP header.
+	TrueClientIP bool
+
+	// XRealIP trusts the X-Real-IP header.
+	XRealIP bool
+
+	// X-Forwarded-For trusts the X-Forwarded-For and X-Forwarded-Proto
+	// headers.
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Proto
+	XForwardedFor bool
```

Comment:
Since we're just accepting an array of headers from the CLI, why not just have this accept an array of headers as well and remove all of the 4x duplicated logic below? It seems the reason is because we want to prioritize some headers over others, but customers can configure that themselves by specifying the order on the CLI.

Reply:
Agreed agreed I'll adjust